File: backend/routes/shopkeeper.py
from flask import Blueprint, request, jsonify
from models import db, User, Shop, Product, Order
import jwt
import os

shopkeeper_bp = Blueprint('shopkeeper', __name__)


# ===========================================
# GET CURRENT USER
# ===========================================

def get_current_user():
    token = request.headers.get("Authorization", "").replace("Bearer ", "")

    if not token:
        return None

    try:
        payload = jwt.decode(
            token,
            os.getenv("SECRET_KEY", "fallback-secret-key-123"),
            algorithms=["HS256"]
        )

        return db.session.get(User, payload["user_id"])

    except jwt.PyJWTError:
        return None


# ===========================================
# DASHBOARD
# ===========================================

@shopkeeper_bp.route("/dashboard", methods=["GET"])
def dashboard():

    user = get_current_user()

    if not user:
        return jsonify({"error": "Unauthorized"}), 401

    shop = db.session.scalars(
        db.select(Shop).filter_by(owner_id=user.id)
    ).first()

    if not shop:
        return jsonify({"error": "Shop not found"}), 404

    products = db.session.scalars(
        db.select(Product).filter_by(shop_id=shop.id)
    ).all()

    # -----------------------------------
    # TOTAL STOCK
    # Example:
    # 199 + 100 + 50 = 349
    # -----------------------------------

    total_stock = sum(product.quantity for product in products)

    # -----------------------------------
    # TOTAL PRODUCT PRICE
    # Example:
    # 50 + 50 + 35.5 = 135.5
    # -----------------------------------

    total_product_price = round(
        sum(product.price for product in products),
        2
    )

    # -----------------------------------
    # INVENTORY VALUE
    # Example:
    # (199×50)+(100×50)+(50×35.5)=16725
    # -----------------------------------

    inventory_value = round(
        sum(product.price * product.quantity for product in products),
        2
    )

    # -----------------------------------
    # LOW STOCK ITEMS
    # -----------------------------------

    low_stock_count = sum(
        1
        for product in products
        if product.quantity <= 5
    )

    return jsonify({

        "total_stock": total_stock,

        "total_product_price": total_product_price,

        "inventory_value": inventory_value,

        "low_stock_count": low_stock_count

    }), 200


# ===========================================
# GET PRODUCTS
# ===========================================

@shopkeeper_bp.route("/products", methods=["GET"])
def get_shopkeeper_products():

    user = get_current_user()

    if not user:
        return jsonify({"error": "Unauthorized"}), 401

    shop = db.session.scalars(
        db.select(Shop).filter_by(owner_id=user.id)
    ).first()

    if not shop:
        return jsonify({"error": "Shop not found"}), 404

    products = db.session.scalars(
        db.select(Product).filter_by(shop_id=shop.id)
    ).all()

    return jsonify([

        {
            "id": product.id,
            "name": product.name,
            "category": product.category,
            "price": product.price,
            "quantity": product.quantity
        }

        for product in products

    ]), 200


# ===========================================
# ADD PRODUCT
# ===========================================

@shopkeeper_bp.route("/products", methods=["POST"])
def add_product():

    user = get_current_user()

    if not user:
        return jsonify({"error": "Unauthorized"}), 401

    shop = db.session.scalars(
        db.select(Shop).filter_by(owner_id=user.id)
    ).first()

    if not shop:
        return jsonify({"error": "Shop not found"}), 404

    data = request.get_json()

    if not data:
        return jsonify({"error": "Invalid data"}), 400

    product = Product(

        shop_id=shop.id,

        name=data["name"],

        category=data["category"],

        price=float(data["price"]),

        quantity=int(data["quantity"])

    )

    db.session.add(product)

    db.session.commit()

    return jsonify({

        "message": "Product added successfully",

        "id": product.id

    }), 201


# ===========================================
# DELETE PRODUCT
# ===========================================

@shopkeeper_bp.route("/products/<int:id>", methods=["DELETE"])
def delete_product(id):

    user = get_current_user()

    if not user:
        return jsonify({"error": "Unauthorized"}), 401

    shop = db.session.scalars(
        db.select(Shop).filter_by(owner_id=user.id)
    ).first()

    if not shop:
        return jsonify({"error": "Shop not found"}), 404

    product = db.session.get(Product, id)

    if not product:
        return jsonify({"error": "Product not found"}), 404

    if product.shop_id != shop.id:
        return jsonify({"error": "Unauthorized"}), 403

    db.session.delete(product)

    db.session.commit()

    return jsonify({

        "message": "Deleted"

    }), 200


# ===========================================
# GET ORDERS
# ===========================================

@shopkeeper_bp.route("/orders", methods=["GET"])
def get_orders():

    user = get_current_user()

    if not user:
        return jsonify({"error": "Unauthorized"}), 401

    shop = db.session.scalars(
        db.select(Shop).filter_by(owner_id=user.id)
    ).first()

    if not shop:
        return jsonify({"error": "Shop not found"}), 404

    orders = db.session.scalars(
        db.select(Order)
        .filter_by(shop_id=shop.id)
        .order_by(Order.created_at.desc())
    ).all()

    order_list = []

    for order in orders:

        items = []

        for item in order.items:

            items.append({
                "product": item.product_ref.name,
                "quantity": item.quantity,
                "price": item.price
            })

        order_list.append({

            "id": order.id,

            "customer_name": order.customer.name,

            "shop_name": shop.name,

            "status": order.status,

            "payment_status": order.payment_status,

            "total_price": order.total_price,

            "created_at": order.created_at.strftime("%d-%m-%Y %I:%M %p"),

            "items": items

        })

    return jsonify(order_list), 200

# ===========================================
# MARK DELIVERED
# ===========================================

@shopkeeper_bp.route("/orders/<int:id>/deliver", methods=["POST"])
def mark_delivered(id):

    user = get_current_user()

    if not user:
        return jsonify({"error": "Unauthorized"}), 401

    order = db.session.get(Order, id)

    if not order:
        return jsonify({"error": "Order not found"}), 404

    order.status = "Delivered"

    db.session.commit()

    return jsonify({

        "message": "Order marked as delivered"

    }), 200


# ===========================================
# LOCATION
# ===========================================

@shopkeeper_bp.route("/location", methods=["POST"])
def update_shop_location():

    return jsonify({

        "message": "Location updates disabled"

    }), 200

File: frontend/shopkeeper/shopkeeper.js
let currentUser = JSON.parse(localStorage.getItem("currentUser"));

// ==============================
// AUTH HEADERS
// ==============================

function authHeaders() {
    return {
        "Authorization": `Bearer ${localStorage.getItem("token")}`,
        "Content-Type": "application/json"
    };
}

// ==============================
// ALERT
// ==============================

function showAlert(message, type) {
    alert(message);
    console.log(`${type}: ${message}`);
}

// ==============================
// INITIAL LOAD
// ==============================

window.addEventListener("DOMContentLoaded", () => {

    if (!currentUser || currentUser.role !== "shopkeeper") {
        window.location.href = "../auth/auth.html";
        return;
    }

    if (!localStorage.getItem("token")) {
        window.location.href = "../auth/auth.html";
        return;
    }

    document.getElementById("userName").textContent = currentUser.name;

    loadDashboard();
    switchSection(null, "dashboard");

});

// ==============================
// SECTION SWITCH
// ==============================

function switchSection(event, sectionName) {

    document.querySelectorAll(".section").forEach(section => {
        section.classList.remove("active");
    });

    document
        .getElementById(`section-${sectionName}`)
        .classList.add("active");

    document.querySelectorAll(".sidebar-btn").forEach(btn => {
        btn.classList.remove("active");
    });

    if (event) {
        event.currentTarget.classList.add("active");
    }

    if (sectionName === "dashboard") {
        loadDashboard();
    }

    if (sectionName === "inventory") {
        loadInventory();
    }

    if (sectionName === "orders-shop") {
        loadShopOrders();
    }

}

// ==============================
// DASHBOARD
// ==============================

async function loadDashboard() {

    try {

        const response = await fetch(
            `${API_BASE}/shopkeeper/dashboard`,
            {
                headers: authHeaders()
            }
        );

        if (response.status === 401) {
            handleLogout();
            return;
        }

        const data = await response.json();

        document.getElementById("totalProducts").textContent =
            data.total_stock ?? 0;

        document.getElementById("inventoryValue").textContent =
            "₹" + (data.inventory_value ?? 0);

        document.getElementById("revenueToday").textContent =
            "₹" + (data.total_product_price ?? 0);

        document.getElementById("lowStockCount").textContent =
            data.low_stock_count ?? 0;

    }

    catch (error) {

        console.error(error);

        showAlert(
            "Unable to load dashboard.",
            "danger"
        );

    }

}

// ==============================
// INVENTORY
// ==============================

async function loadInventory() {

    try {

        const response = await fetch(
            `${API_BASE}/shopkeeper/products`,
            {
                headers: authHeaders()
            }
        );

        if (response.status === 401) {
            handleLogout();
            return;
        }

        const products = await response.json();

        const container =
            document.getElementById("inventoryTable");

        if (!Array.isArray(products) || products.length === 0) {

            container.innerHTML =
                "<p>No products found. Add your first product!</p>";

            return;

        }

        container.innerHTML = `

<table class="product-table">

<thead>

<tr>

<th>Name</th>
<th>Category</th>
<th>Price</th>
<th>Quantity</th>
<th>Action</th>

</tr>

</thead>

<tbody>

${products.map(product => `

<tr>

<td>${product.name}</td>

<td>${product.category}</td>

<td>₹${product.price}</td>

<td>${product.quantity}</td>

<td>

<button
class="btn btn-danger"
onclick="deleteProduct('${product.id}')">

Delete

</button>

</td>

</tr>

`).join("")}

</tbody>

</table>

`;

    }

    catch (error) {

        console.error(error);

        showAlert(
            "Failed to load inventory.",
            "danger"
        );

    }

}

// ==============================
// OPEN MODAL
// ==============================

function openAddProductModal() {

    document
        .getElementById("addProductModal")
        .classList.add("active");

}

// ==============================
// CLOSE MODAL
// ==============================

function closeModal(id) {

    document
        .getElementById(id)
        .classList.remove("active");

}
// ==============================
// ADD PRODUCT
// ==============================

async function addProduct() {

    const nameInput = document.getElementById("productName");
    const categoryInput = document.getElementById("productCategory");
    const priceInput = document.getElementById("productPrice");
    const quantityInput = document.getElementById("productQty");

    const name = nameInput.value.trim();
    const category = categoryInput.value.trim();
    const price = parseFloat(priceInput.value);
    const quantity = parseInt(quantityInput.value);

    if (
        !name ||
        !category ||
        isNaN(price) ||
        isNaN(quantity)
    ) {
        showAlert("Please fill all fields.", "danger");
        return;
    }

    try {

        const response = await fetch(
            `${API_BASE}/shopkeeper/products`,
            {
                method: "POST",
                headers: authHeaders(),
                body: JSON.stringify({
                    name,
                    category,
                    price,
                    quantity
                })
            }
        );

        if (!response.ok) {
            throw new Error("Unable to add product.");
        }

        nameInput.value = "";
        categoryInput.value = "";
        priceInput.value = "";
        quantityInput.value = "";

        closeModal("addProductModal");

        // Refresh everything
        await loadInventory();
        await loadDashboard();

        showAlert(
            "Product added successfully.",
            "success"
        );

    }

    catch (error) {

        console.error(error);

        showAlert(
            "Failed to add product.",
            "danger"
        );

    }

}

// ==============================
// DELETE PRODUCT
// ==============================

async function deleteProduct(id) {

    if (!confirm("Delete this product?")) {
        return;
    }

    try {

        const response = await fetch(
            `${API_BASE}/shopkeeper/products/${id}`,
            {
                method: "DELETE",
                headers: authHeaders()
            }
        );

        if (!response.ok) {

            const err = await response.json();

            throw new Error(
                err.error || "Delete failed"
            );

        }

        showAlert(
            "Product deleted successfully.",
            "success"
        );

        // Refresh inventory and dashboard
        await loadInventory();
        await loadDashboard();

    }

    catch (error) {

        console.error(error);

        showAlert(
            error.message,
            "danger"
        );

    }

}

// ==============================
// LOAD ORDERS
// ==============================
async function loadShopOrders() {

    try {

        const response = await fetch(
            `${API_BASE}/shopkeeper/orders`,
            {
                headers: authHeaders()
            }
        );

        const orders = await response.json();

        const container =
            document.getElementById("shopOrdersContainer");

        if (!orders.length) {

            container.innerHTML = `
                <div class="card">
                    <h3>No Orders Yet</h3>
                </div>
            `;

            return;
        }

        container.innerHTML = orders.map((order, index) =>  `

<div class="card order-card">

<h3>Order #${index + 1}</h3>

<p><strong>Customer :</strong> ${order.customer_name}</p>

<p><strong>Shop :</strong> ${order.shop_name}</p>

<p><strong>Status :</strong>
<span class="${order.status.toLowerCase()}">
${order.status}
</span>
</p>

<p><strong>Total :</strong> ₹${order.total_price}</p>

<p><strong>Date :</strong> ${order.created_at}</p>

<h4>Items</h4>

<ul>

${order.items.map(item => `

<li>

${item.product}

×

${item.quantity}

(₹${item.price})

</li>

`).join("")}

</ul>

${order.status.toLowerCase() !== "delivered"

?

`<button
class="btn btn-primary"
onclick="markDelivered(${order.id})">

Mark Delivered

</button>`

:

`<button
class="btn btn-success"
disabled>

Delivered ✓

</button>`

}

</div>

`).join("");

    }

    catch(error){

        console.log(error);

    }

}
// ==============================
// MARK DELIVERED
// ==============================

async function markDelivered(id) {

    try {

        const response = await fetch(
            `${API_BASE}/shopkeeper/orders/${id}/deliver`,
            {
                method: "POST",
                headers: authHeaders()
            }
        );

        if (!response.ok) {
            throw new Error();
        }

        showAlert(
            "Order marked delivered.",
            "success"
        );

        loadShopOrders();

    }

    catch {

        showAlert(
            "Unable to update order.",
            "danger"
        );

    }

}

// ==============================
// LOGOUT
// ==============================

function handleLogout() {

    localStorage.removeItem("currentUser");
    localStorage.removeItem("token");

    window.location.href =
        "../auth/auth.html";

}

// ==============================
// EXPORT FUNCTIONS
// ==============================

window.switchSection = switchSection;
window.handleLogout = handleLogout;
window.openAddProductModal = openAddProductModal;
window.closeModal = closeModal;
window.addProduct = addProduct;
window.deleteProduct = deleteProduct;
window.markDelivered = markDelivered;